transformer_engine/pytorch/csrc/extensions/permutation.cu

```diff
@@ -0,0 +1,245 @@
+                                            row_id_map_ptr, prob_ptr, num_tokens, num_topK,
+                                            num_cols, 0, stream);
+
+      break;
+    }
+    case at::ScalarType::BFloat16: {
+      moe_permutation_launcher<__nv_bfloat16, false>(input_ptr, unpermuted_output_ptr, nullptr,
+                                                     row_id_map_ptr, prob_ptr, num_tokens, num_topK,
+                                                     num_cols, 0, stream);
+
+      break;
+    }
+    case at::ScalarType::Float8_e5m2: {
+      moe_permutation_launcher<__nv_fp8_e5m2, false>(input_ptr, unpermuted_output_ptr, nullptr,
+                                                     row_id_map_ptr, prob_ptr, num_tokens, num_topK,
+                                                     num_cols, 0, stream);
+
+      break;
+    }
+    case at::ScalarType::Float8_e4m3fn: {
+      moe_permutation_launcher<__nv_fp8_e4m3, false>(input_ptr, unpermuted_output_ptr, nullptr,
+                                                     row_id_map_ptr, prob_ptr, num_tokens, num_topK,
+                                                     num_cols, 0, stream);
+
+      break;
+    }
+    default:
+      throw std::runtime_error("Wrong activation tensor type.");
+  }
+
+  return unpermuted_output;
+}
+
+std::tuple<Tensor, Tensor> moe_unpermute_bwd(Tensor input_bwd, Tensor input_fwd, Tensor row_id_map,
+                                             Tensor prob) {
+  const int num_topK = (prob.numel() > 0) ? prob.size(1) : 1;
+  const int num_tokens = (prob.numel() > 0) ? prob.size(0) : row_id_map.size(0);
+  const int num_cols = input_bwd.size(1);
+
+  int *row_id_map_ptr = reinterpret_cast<int *>(getDataPtr(row_id_map, 0));
+  float *prob_ptr = (prob.numel() > 0) ? reinterpret_cast<float *>(getDataPtr(prob, 0)) : nullptr;
+
+  // activations type
+  const at::ScalarType _st = input_bwd.scalar_type();
+
+  // Output buffer alloc
+  Tensor act_grad = torch::empty({input_fwd.size(0), num_cols},
+                                 torch::dtype(_st).device(torch::kCUDA).requires_grad(false));
+  Tensor prob_grad =
+      torch::empty({num_tokens, num_topK},
+                   torch::dtype(torch::kFloat32).device(torch::kCUDA).requires_grad(false));
+  float *prob_grad_ptr = reinterpret_cast<float *>(getDataPtr(prob_grad, 0));
+
+  auto stream = at::cuda::getCurrentCUDAStream().stream();
+
+  void *input_bwd_ptr = getDataPtr(input_bwd, 0);
+  void *input_fwd_ptr = getDataPtr(input_fwd, 0);
+  void *act_grad_ptr = getDataPtr(act_grad, 0);
+
+  switch (_st) {
+    case at::ScalarType::Float: {
+      moe_permutation_launcher<float, true>(input_bwd_ptr, act_grad_ptr, nullptr, row_id_map_ptr,
+                                            prob_ptr, num_tokens, num_topK, num_cols, 0, stream,
+                                            prob_grad_ptr, input_fwd_ptr);
+
+      break;
+    }
+    case at::ScalarType::Half: {
```

Comment:
Hi,
We have `TRANSFORMER_ENGINE_TYPE_SWITCH_INPUT` macro defined in `transformer_engine/common/ common.h` which converts torch dtypes to cuda dtypes. Could you consider if it is possible to use it?

If we could then we don't need to template the dtype, just need to have a bool to indicate permute/unpermute.

Reply:
Hi. I think we can template functions in this file even further. 
L68 - L112 block can be reused for all three functions, and thus can be moved to a separate function, let's say `moe_permutation_launcher`. The current `moe_permutation_launcher` should be renamed to `nvte_permutation` as mentioned in another comment.

Reply:
Done.